gh-5500: Support various Authenticated Request methods

mhyeon-lee opened the pull request:
To support RFC6750 Section 2 Authenticated Requests Specifications
https://tools.ietf.org/html/rfc6750#section-2

gh-5500

mhyeon-lee commented:
Tests that did not involve changes in this PR build failed. 
It seems that one of the SNAPSHOT libraries that depend on is breaking change.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:
```diff
@@ -64,6 +66,16 @@ public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2Authentic
 			);
 			throw new OAuth2AuthenticationException(oauth2Error, oauth2Error.toString());
 		}
+		UserInfoRequestSchema userInfoRequestSchema = userRequest.getClientRegistration().getProviderDetails().getUserInfoEndpoint().getUserInfoRequestSchema();
```
If the user has not configured `authenticationMethod` we should default to `HEADER`. So this check would not be needed.

jgrandja commented on oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/UserInfoRequestSchema.java:
```diff
@@ -0,0 +1,83 @@
+ *  Clients MUST NOT use more than one method to transmit the token in each request.
+ *
+ * <p>
+ * The OAuth 2.0 Authorization Framework defines three standard request schema types:
```
I believe you meant `scheme` instead of `schema`? 

As per spec,

> 2.  **Authenticated** Requests

   This section defines three **methods**...

I believe the name `AuthenticationMethod` aligns best with the spec language.

jgrandja commented on oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/UserInfoRequestSchema.java:
```diff
@@ -0,0 +1,83 @@
+public final class UserInfoRequestSchema implements Serializable {
+	private static final long serialVersionUID = SpringSecurityCoreVersion.SERIAL_VERSION_UID;
+	public static final UserInfoRequestSchema HEADER = new UserInfoRequestSchema("header");
+	public static final UserInfoRequestSchema FORM = new UserInfoRequestSchema("form");
```
Can you also add `QUERY` but don't implement/use it in the `OAuth2UserService`. This will come in handy in Resource Server feature.

mhyeon-lee commented:
@jgrandja 
I modified it to your comment.
There are a few additional considerations.

1. AuthenticationMethod is HEADER by default but intentionally needs to be checked to prevent null setting.
Can we validate for build on ClientRegistration?

2. I added QUERY type to AuthenticationMethod, but I did not implement it in OAuth2UserService as you mentioned. 
Currently, if you set the AuthenticationMethod to QUERY, it will behave like HEADER.

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/userinfo/DefaultReactiveOAuth2UserServiceTests.java:
```diff
@@ -140,6 +145,51 @@ public void loadUserWhenUserInfoSuccessResponseThenReturnUser() throws Exception
 
+	// gh-5500
+	@Test
+	public void loadUserWhenAuthenticationMethodHeaderSuccessResponseThenHttpMethod() throws Exception {
```
Can you rename to ...ThenHttpMethod**Get**

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/userinfo/DefaultReactiveOAuth2UserServiceTests.java:
```diff
@@ -140,6 +145,51 @@ public void loadUserWhenUserInfoSuccessResponseThenReturnUser() throws Exception
+
+	// gh-5500
+	@Test
+	public void loadUserWhenAuthenticationMethodFormSuccessResponseThenHttpMethod() throws Exception {
```
Can you rename to ...ThenHttpMethod**Post**

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/userinfo/DefaultReactiveOAuth2UserServiceTests.java:
```diff
@@ -140,6 +145,51 @@ public void loadUserWhenUserInfoSuccessResponseThenReturnUser() throws Exception
+		assertThat(request.getMethod()).isEqualTo(HttpMethod.POST.name());
+		assertThat(request.getHeader(HttpHeaders.ACCEPT)).isEqualTo(MediaType.APPLICATION_JSON_VALUE);
+		assertThat(request.getHeader(HttpHeaders.CONTENT_TYPE)).contains(MediaType.APPLICATION_FORM_URLENCODED_VALUE);
+		assertThat(request.getUtf8Body()).isEqualTo("access_token=" + this.accessToken.getTokenValue());
```
`getUtf8Body()` is deprecated, please choose another

jgrandja commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserServiceTests.java:
```diff
@@ -237,4 +246,73 @@ public void loadUserWhenUserInfoSuccessResponseThenAcceptHeaderJson() throws Exc
+		assertThat(request.getMethod()).isEqualTo(HttpMethod.POST.name());
+		assertThat(request.getHeader(HttpHeaders.ACCEPT)).isEqualTo(MediaType.APPLICATION_JSON_VALUE);
+		assertThat(request.getHeader(HttpHeaders.CONTENT_TYPE)).contains(MediaType.APPLICATION_FORM_URLENCODED_VALUE);
+		assertThat(request.getUtf8Body()).isEqualTo("access_token=" + this.accessToken.getTokenValue());
```
`getUtf8Body()` is deprecated, please choose another

jgrandja commented:
In response to your comment:
> AuthenticationMethod is HEADER by default but intentionally needs to be checked to prevent null setting.
Can we validate for build on ClientRegistration?

I don't feel we need to validate it. The NPE can only happen if the user explicitly calls `authenticationMethod(null)` on the `Builder` which IMO would be a programming error. 
Also, we don't validate on `ClientAuthenticationMethod`

mhyeon-lee commented:
@jgrandja 
I modified everything you commented on.
Do you have any further changes?

jgrandja commented:
Great job @mhyeon-lee! This is now in master.

Thanks so much for all your contributions and feedback as of late! We really appreciate it.
Looking forward to some more :)

mhyeon-lee commented:
@jgrandja 
Thank you very much.
It was a feature I needed. v.5.1.0 is expected.
